File: src/lib/aethos/astrology/metadata.ts
import { createHash } from "node:crypto";
import type { CalculationMetadataV2, CalculationMode, HouseSystem, ZodiacMode } from "./types";

export function createCalculationId(prefix = "calc"): string {
  return `${prefix}-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
}

export function createInputHash(input: unknown): string {
  return createHash("sha256").update(JSON.stringify(sortObject(input))).digest("hex");
}

function sortObject(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(sortObject);
  if (value && typeof value === "object") {
    return Object.keys(value as Record<string, unknown>)
      .sort()
      .reduce<Record<string, unknown>>((acc, key) => {
        acc[key] = sortObject((value as Record<string, unknown>)[key]);
        return acc;
      }, {});
  }
  return value;
}

export function attachCalculationMetadata(input: {
  providerId: string;
  providerVersion: string;
  calculationMode: CalculationMode;
  sourceInput: unknown;
  timezone?: string;
  coordinates?: { latitude: number; longitude: number };
  houseSystem?: HouseSystem;
  zodiacMode?: ZodiacMode;
  ephemerisSource?: string;
  warnings?: string[];
}): CalculationMetadataV2 {
  return {
    calculationId: createCalculationId("aethos"),
    providerId: input.providerId,
    providerVersion: input.providerVersion,
    calculationMode: input.calculationMode,
    generatedAt: new Date().toISOString(),
    inputHash: createInputHash(input.sourceInput),
    timezone: input.timezone ?? "unknown",
    coordinates: input.coordinates,
    houseSystem: input.houseSystem,
    zodiacMode: input.zodiacMode,
    ephemerisSource: input.ephemerisSource,
    warnings: input.warnings ?? []
  };
}

export function validateCalculationMetadata(metadata: CalculationMetadataV2): boolean {
  return Boolean(
    metadata.calculationId &&
      metadata.providerId &&
      metadata.providerVersion &&
      metadata.calculationMode &&
      metadata.generatedAt &&
      metadata.inputHash
  );
}

export function compareCalculationInputs(a: unknown, b: unknown): boolean {
  return createInputHash(a) === createInputHash(b);
}

File: services/calculation/app/metadata.py
from datetime import datetime, timezone
from hashlib import sha256
from uuid import uuid4
import json
from app.config import get_settings
from app.models.astrology import CalculationMode, Coordinates, HouseSystem, ZodiacMode
from app.models.metadata import CalculationMetadata


def canonical_json(payload: object) -> str:
    return json.dumps(payload, sort_keys=True, separators=(",", ":"), default=str)


def create_input_hash(payload: object) -> str:
    return sha256(canonical_json(payload).encode("utf-8")).hexdigest()


def create_metadata(
    *,
    provider_id: str,
    provider_version: str,
    calculation_mode: CalculationMode,
    source_input: object,
    timezone_name: str,
    normalized_utc: str | None = None,
    coordinates: Coordinates | None = None,
    house_system: HouseSystem | None = None,
    zodiac_mode: ZodiacMode | None = None,
    ephemeris_source: str | None = None,
    warnings: list[str] | None = None,
) -> CalculationMetadata:
    return CalculationMetadata(
        calculationId=f"calc-{uuid4()}",
        providerId=provider_id,
        providerVersion=provider_version,
        serviceVersion=get_settings().service_version,
        generatedAt=datetime.now(timezone.utc).isoformat(),
        inputHash=create_input_hash(source_input),
        timezone=timezone_name,
        normalizedUtc=normalized_utc,
        coordinates=coordinates,
        houseSystem=house_system,
        zodiacMode=zodiac_mode,
        calculationMode=calculation_mode,
        ephemerisSource=ephemeris_source,
        warnings=warnings or [],
    )
